kuma/static/js/newsletter.js

```diff
@@ -0,0 +1,221 @@
+;(function(doc) {
+    'use strict';
+
+    // !! this file assumes only one signup form per page !!
+
+    var newsletterForm = doc.getElementById('newsletterForm');
+    var isHidden = false;
+    var isArticle = false;
+
+    if(!newsletterForm) {
+        return;
+    } else {
+        // check for local storage
+        if(mdn.features.localStorage) {
+            // if hidden by local storage
+            isHidden = localStorage.getItem('newsletterHide') == 'true' ? true : false;
+        }
+
+        // check for article
+        if(doc.querySelector('.wiki-block-newsletter')) {
+            isArticle = true;
+        }
+
+        if(!mdn.features.localStorage || !isArticle) {
+            newsletter();
+        } else if(isArticle && !isHidden) {
+            newsletter();
+            newsletterAddHideButton();
+        } else if (isArticle && isHidden) {
+            newsletterHide();
+        }
+    }
+
+    function newsletter() {
+
+        var newsletterEmailInput = doc.getElementById('newsletterEmailInput');
+        var newsletterPrivacy = doc.getElementById('newsletterPrivacy');
+
+        // handle errors
+        var errorArray = [];
+        var newsletterErrors = doc.getElementById('newsletterErrors');
+        function newsletterError(event) {
+            var errorList = doc.createElement('ul');
+            errorList.className ='errorlist';
+
+            if(errorArray.length) {
+                for (var i = 0; i < errorArray.length; i++) {
+                    var item = doc.createElement('li');
+                    item.appendChild(doc.createTextNode(errorArray[i]));
+                    errorList.appendChild(item);
+                }
+            } else {
+                // no error messages, forward to server for better troubleshooting
+                newsletterForm.setAttribute('data-skip-xhr', true);
+                newsletterForm.submit();
+            }
+
+            newsletterErrors.appendChild(errorList);
+            newsletterErrors.style.display = 'block';
+
+            mdn.analytics.trackEvent({
+                'category': 'newsletter',
+                'action': 'progression',
+                'label': 'error'
+            });
```

Comment:
Oh, probably including this? Perhaps event factor the lines out into a small function like:

```
function showFormErrors(errorList) {
    newsletterErrors.appendChild(errorList);
    newsletterErrors.style.display = 'block';

    mdn.analytics.trackEvent({
        'category': 'newsletter',
        'action': 'progression',
        'label': 'error'
    });
}
```